gettsim/taxes/zve.py

```diff
@@ -312,58 +316,151 @@ def calc_gross_e7(tax_unit, params):
+        item_3 = np.minimum(0.5 * (vorsorg_rest - item_2), 0.5 * params["grundbet"])
+    # For the married couple, the same stuff, but with tu totals.
+    if tax_unit["zveranl"].max():
+        for var in ["m_wage", "rvbeit", "gkvbeit"]:
+            tax_unit[f"{var}_tu"] = tax_unit.loc[~tax_unit["child"], "m_wage"].sum()
+        if params["year"] <= 2019:
+            item_1 = 0.5 * np.maximum(
+                2 * params["vorwegab"]
+                - params["kuerzquo"] * 12 * tax_unit["m_wage_tu"],
+                0,
+            )
+        else:
+            item_1 = 0
+
+        vorsorg_rest = 0.5 * np.maximum(
+            12 * (tax_unit["rvbeit_tu"] + tax_unit["gkvbeit_tu"]) - item_1, 0
+        )
+        item_2 = 0.5 * np.minimum(params["grundbet"], vorsorg_rest)
+        item_3 = 0.5 * np.minimum((vorsorg_rest - item_2), 2 * params["grundbet"])
+
+    # Finally, add up all three amounts and assign it to the adults.
+    vorsorge = ~tax_unit["child"] * (item_1 + item_2 + item_3).astype(int)
+    # The test requires a named item. If there is only one value, create a list.
+    return vorsorge
+
+
+def vorsorge_since_2005(tax_unit, params, soz_vers_beitr_params):
+    """ With the 2005 reform, no taxpayer was supposed to be affected negatively.
+        Therefore, one needs to compute amounts
+        (2004 and 2005 regime) and take the higher one.
+    """
+    vors2004 = vorsorge_pre_2005(tax_unit, params, soz_vers_beitr_params)
+    vors2005 = _vorsorge_since_2005(tax_unit, params, soz_vers_beitr_params)
+
+    return np.maximum(vors2004, vors2005)
 
-def vorsorge_dummy(tax_unit, params, soz_vers_beitr_params):
-    return 0
+
+def vorsorge_since_2010(tax_unit, params, soz_vers_beitr_params):
+    """ After a supreme court ruling, the 2005 rule had to be changed in 2010.
+        Therefore, one needs to compute amounts
+        (2004 and 2010 regime) and take the higher one. (§10 (3a) EStG).
+        Sidenote: The 2010 ruling is by construction
+        *always* more or equally beneficial than the 2005 one,
+        so no need for a separate check there.
+    """
+    vors2004 = vorsorge_pre_2005(tax_unit, params, soz_vers_beitr_params)
+    vors2010 = _vorsorge_since_2010(tax_unit, params, soz_vers_beitr_params)
+
+    return np.maximum(vors2004, vors2010)
+
+
+def calc_altersvors_aufwend(tax_unit, params):
+    """ Calculates deductible old-age contributions since 2005
+    """
+    altersvors = ~tax_unit["child"] * (
+        vorsorge_year_faktor(params["year"])
+        * (12 * 2 * tax_unit["rvbeit"] + (12 * tax_unit["priv_pens_contr"]))
+        - (12 * tax_unit["rvbeit"])
+    ).astype(int)
+
+    return np.minimum(params["vorsorg_rv_max"], altersvors)
+
+
+def vorsorge_year_faktor(year):
+    """ in the calculation of *Vorsorgeaufwendungen*,
+    there is year-dependent factor to be calculated.
+
+    year: int
+
+    returns the factor
+    """
+    return 0.6 + 0.02 * (min(year, 2025) - 2005)
```

Comment:
Is there a better solution than hardcoding?

Reply:
I don't know. The alternative would be to specify the values year-wise in the param file.

Reply:
While it looks unusual, this is not "harder coded" than any other tax tariff of benefit amount calculation.

Reply:
This seems indeed special enough so as not to move it to the parameters files.

A whole function seems too much to me, though. I would just do:

```python
einführungsfaktor = 0.6 + 0.02 * (min(year, 2025) - 2005)
```

directly in   `calc_altersvors_aufwend`. And we should add to the docstring of that function that it includes the phasing in, explictly cautioning against using it for years prior to 2005 (who knows whether somebody wants to to a policy change of the type 'what if the current system had been around in 1990'?)